src/computability/turing_machine.lean

```diff
@@ -1309,74 +1774,51 @@ parameters {σ : Type*} [inhabited σ]
 local notation `stmt₂` := TM2.stmt Γ Λ σ
 local notation `cfg₂` := TM2.cfg Γ Λ σ
 
-inductive stackel (k : K)
-| val : Γ k → stackel
-| bottom [] : stackel
-| top [] : stackel
-
-instance stackel.inhabited (k) : inhabited (stackel k) :=
-⟨stackel.top _⟩
-
-def stackel.is_bottom {k} : stackel k → bool
-| (stackel.bottom _) := tt
-| _ := ff
-
-def stackel.is_top {k} : stackel k → bool
-| (stackel.top _) := tt
-| _ := ff
-
-def stackel.get {k} : stackel k → option (Γ k)
-| (stackel.val a) := some a
-| _ := none
-
-section
-open stackel
-
-def stackel_equiv {k} : stackel k ≃ option (option (Γ k)) :=
-begin
-  refine ⟨λ s, _, λ s, _, _, _⟩,
-  { cases s, exacts [some (some s), none, some none] },
-  { rcases s with _|_|s, exacts [bottom _, top _, val s] },
-  { intro s, cases s; refl },
-  { intro s, rcases s with _|_|s; refl },
-end
-
-end
-
-def Γ' := ∀ k, stackel k
+/-- The alphabet of the TM2 simulator on TM1 is a marker for the stack bottom,
+plus a vector of stack elements for each stack, or none if the stack does not extend this far. -/
+@[nolint unused_arguments]
+def Γ' := bool × ∀ k, option (Γ k)
 
-instance Γ'.inhabited : inhabited Γ' := ⟨λ _, default _⟩
-
-instance stackel.fintype {k} [fintype (Γ k)] : fintype (stackel k) :=
-fintype.of_equiv _ stackel_equiv.symm
+instance Γ'.inhabited : inhabited Γ' := ⟨⟨ff, λ _, none⟩⟩
 
 instance Γ'.fintype [fintype K] [∀ k, fintype (Γ k)] : fintype Γ' :=
-pi.fintype
+prod.fintype _ _
 
+/-- A stack action is a command that interacts with the top of a stack. Our default position
+is at the bottom of all the stacks, so we have to hold on to this action while going to the end
+to modify the stack. -/
 inductive st_act (k : K)
 | push : (σ → Γ k) → st_act
-| pop : bool → (σ → option (Γ k) → σ) → st_act
+| peek : (σ → option (Γ k) → σ) → st_act
+| pop : (σ → option (Γ k) → σ) → st_act
+
+instance st_act.inhabited {k} : inhabited (st_act k) := ⟨st_act.peek (λ s _, s)⟩
 
 section
 open st_act
 
-instance st_act.inhabited {k} : inhabited (st_act k) :=
-⟨pop (default _) (λ s _, s)⟩
-
+/-- The TM2 statement corresponding to a stack action. -/
+@[nolint unused_arguments]
```

Comment:
Please avoid `nolint`.  In particular, please don't add unjustified `nolint` attributes for declarations that are already in `nolints.txt`.  If you are renaming declarations, you are allowed to rename them in `nolints.txt` as well (this is the "grandfathering exception").

Reply:
My understanding is that a nolint in `nolints.txt` means the theorem/definition is grandfathered in and whether or not the lint is justified has not been evaluated, while a `@[nolint]` means that the lint *has* been evaluated and found to be inapplicable. But I will add comments on all the nolints to make this clear. (Won't the comments interfere with the doc comment?)

Reply:
Yes, please document why.  You can have as many regular comments as you want between the docstring and the declaration.